<a href="https://colab.research.google.com/github/Joel-Alejandro-DM/DrivenPath/blob/main/L02_03_Neurona_logistica_Diabetes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Neurona Logística
## Dr. Carlos Villaseñor

Paso 1. De la carpeta "Datasets" del repositorio descarga alguno de los siguientes datasets: "cancer.csv" o "diabetes.csv"

In [332]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
!wget https://raw.githubusercontent.com/Dr-Carlos-Villasenor/Clase_Aprendizaje_Profundo/refs/heads/main/Datasets/diabetes.csv

--2025-08-26 04:50:18--  https://raw.githubusercontent.com/Dr-Carlos-Villasenor/Clase_Aprendizaje_Profundo/refs/heads/main/Datasets/diabetes.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 23873 (23K) [text/plain]
Saving to: ‘diabetes.csv.3’

diabetes.csv.3      100%[===================>]  23.31K  --.-KB/s    in 0.003s  

2025-08-26 04:50:18 (7.94 MB/s) - ‘diabetes.csv.3’ saved [23873/23873]



In [333]:
ds = pd.read_csv('diabetes.csv')
Y = ds[['Outcome']].to_numpy().T
X = ds.drop(['Outcome'], axis=1).to_numpy().T
ds.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


Paso 2. Crea una neurona logística que aprenda el dataset tratando de predecir la salida.

In [334]:
class Logistic_Neuron:

    def __init__(self, n_inputs, learning_rate=0.1):
        self.w = - 1 + 2 * np.random.rand(n_inputs)
        self.b = - 1 + 2 * np.random.rand()
        self.eta = learning_rate

    def predict_proba(self, X):
        Z = np.dot(self.w, X) + self.b
        Y_est = 1/(1+np.exp(-Z))
        return Y_est

    def predict(self, X, umbral=0.5):
        Z = np.dot(self.w, X) + self.b
        Y_est = 1/(1+np.exp(-Z))
        return 1 * (Y_est > umbral)

    def train(self, X, Y, epochs=100):
        p = X.shape[1]
        for _ in range(epochs):
            Y_est = self.predict_proba(X)
            self.w += (self.eta/p) * np.dot((Y - Y_est), X.T).ravel()
            self.b += (self.eta/p) * np.sum(Y - Y_est)

Paso 3. Con un umbral de 0.5, calcula el porcentaje de los diagnósticos acertados por la neurona.

In [336]:
from re import I
neuron = Logistic_Neuron(X.shape[0], learning_rate = 0.005)
neuron.train(X, Y, epochs=10000)

Y_predict = neuron.predict(X, 0.5).ravel()
Y = Y.ravel()

ok = 0
for i in range(Y.shape[0]):
  if Y[i] == Y_predict[i]:
    ok += 1

print(f'Porcentaje de aciertos: {100 * ok/Y.shape[0]}',"\b%")

Porcentaje de aciertos: 64.84375 %


Paso 4. Reporta tu código y resultados.

La moda se encuentra al rededor del 65% es el promedio de resultados obtenidos con las epocas y factor de aprendizaje utilizados para esta base data con datos dispersos. casos extremos 33% y 70%